<?xml version="1.0"?>

<robot name="mobrob" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <xacro:include filename="$(find myrobot_model)/urdf/mobrob_gpu.gazebo" />
  <!--<xacro:include filename="$(find myrobot_model)/urdf/mobrob.gazebo" />-->
  <xacro:include filename="$(find myrobot_model)/urdf/materials.urdf.xacro" />

  <link name="base_footprint"></link>

  <joint name="base_link_joint" type="fixed">
    <origin xyz="0 0 0.05" rpy="0 0 0" />
    <parent link="base_footprint"/>
    <child link="base_link" />
  </joint>

  <link name="base_link">
    <visual>
      <geometry>
        <box size="0.2 0.25 0.055"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <box size="0.2 0.25 0.055"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="1.1"/>
      <inertia
        ixx="0.006006" ixy="0.0" ixz="0.0"
        iyy="0.003944" iyz="0.0"
        izz="0.009396"/>
    </inertial>
  </link>

  <link name="upperbase_link">
    <visual>
      <origin xyz="0 0 0.0235" rpy="0 0 0" />
      <geometry>
        <box size="0.4 0.25 0.02"/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0 0 -0.0235" rpy="0 0 0" />
      <geometry>
        <box size="0.4 0.25 0.02"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <box size="0.4 0.25 0.055"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="2.6"/>
      <inertia
        ixx="0.01449" ixy="0.0" ixz="0.0"
        iyy="0.03558" iyz="0.0"
        izz="0.04821"/>
    </inertial>
  </link>
  <joint name="base_to_upperbase" type="fixed">
    <parent link="base_link"/>
    <child link="upperbase_link"/>
    <origin xyz="0 0 0.06"/>
  </joint>

  <link name="laser">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.03" radius="0.04"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.03" radius="0.04"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.2"/>
      <inertia
        ixx="0.0001021" ixy="0.0" ixz="0.0"
        iyy="0.0001021" iyz="0.0"
        izz="0.0001719"/>
    </inertial>
  </link>
  <joint name="base_to_laser" type="fixed">
    <parent link="base_link"/>
    <child link="laser"/>
    <!--<origin rpy="0 0 1.57075" xyz="0.055 0 0.06"/>-->
    <origin rpy="0 0 ${-pi}" xyz="0.055 0 0.06"/>
  </joint>

  <link name="right_front_wheel">
    <visual>
      <origin rpy="1.57075 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.032" radius="0.03"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin rpy="1.57075 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.032" radius="0.03"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="1.57075 0 0" xyz="0 0 0"/>
      <mass value="0.08"/>
      <inertia
        ixx="0.00002611" ixy="0.0" ixz="0.0"
        iyy="0.00003786" iyz="0.0"
        izz="0.00002611"/>
    </inertial>
  </link>
  <joint name="right_front_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="right_front_wheel"/>
    <origin rpy="0 0 0" xyz="0.142 0.109 -0.0175"/>
    <axis xyz="0 1 0"/>
  </joint>

  <link name="right_back_wheel">
    <visual>
      <origin rpy="1.57075 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.032" radius="0.03"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin rpy="1.57075 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.032" radius="0.03"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="1.57075 0 0" xyz="0 0 0"/>
      <mass value="0.08"/>
      <inertia
        ixx="0.00002611" ixy="0.0" ixz="0.0"
        iyy="0.00003786" iyz="0.0"
        izz="0.00002611"/>
    </inertial>
  </link>
  <joint name="right_back_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="right_back_wheel"/>
    <origin rpy="0 0 0" xyz="-0.147 0.109 -0.0175"/>
    <axis xyz="0 1 0"/>
  </joint>

  <link name="left_front_wheel">
    <visual>
      <origin rpy="1.57075 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.032" radius="0.03"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin rpy="1.57075 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.032" radius="0.03"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="1.57075 0 0" xyz="0 0 0"/>
      <mass value="0.08"/>
      <inertia
        ixx="0.00002611" ixy="0.0" ixz="0.0"
        iyy="0.00003786" iyz="0.0"
        izz="0.00002611"/>
    </inertial>
  </link>
  <joint name="left_front_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="left_front_wheel"/>
    <origin rpy="0 0 0" xyz="0.142 -0.109 -0.0175"/>
    <axis xyz="0 1 0"/>
  </joint>

  <link name="left_back_wheel">
    <visual>
      <origin rpy="1.57075 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.032" radius="0.03"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin rpy="1.57075 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.032" radius="0.03"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="1.57075 0 0" xyz="0 0 0"/>
      <mass value="0.08"/>
      <inertia
        ixx="0.00002611" ixy="0.0" ixz="0.0"
        iyy="0.00003786" iyz="0.0"
        izz="0.00002611"/>
    </inertial>
  </link>
  <joint name="left_back_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="left_back_wheel"/>
    <origin rpy="0 0 0" xyz="-0.147 -0.109 -0.0175"/>
    <axis xyz="0 1 0"/>
  </joint>

  <link name="gripper_stand">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0.065"/>
      <geometry>
        <box size="0.03 0.13 0.18"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.065"/>
      <geometry>
        <box size="0.03 0.13 0.18"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.28"/>
      <inertia
        ixx="0.001154" ixy="0.0" ixz="0.0"
        iyy="0.0007792" iyz="0.0"
        izz="0.0004165"/>
    </inertial>
  </link>
  <joint name="base_to_gripper_stand_joint" type="fixed">
    <parent link="base_link"/>
    <child link="gripper_stand"/>
    <origin rpy="0 0 0" xyz="0.2 0 0.1175"/>
  </joint>

  <link name="gripper_base">
    <visual>
      <origin rpy="0 0 0" xyz="-0.04 0 0"/>
      <geometry>
        <box size="0.04 0.25 0.057"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="-0.04 0 0"/>
      <geometry>
        <box size="0.04 0.25 0.057"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.228"/>
      <inertia
        ixx="0.001249" ixy="0.0" ixz="0.0"
        iyy="0.00009213" iyz="0.0"
        izz="0.001218"/>
    </inertial>
  </link>
  <joint name="gripper_stand_to_gripper_joint" type="prismatic">
    <parent link="gripper_stand"/>
    <child link="gripper_base"/>
    <axis xyz="0 0 1"/>
    <limit effort="1000.0" lower="-0.11" upper="0" velocity="0.5"/>
    <origin rpy="0 0 0" xyz="0.075 0 0"/>
  </joint>

  <link name="right_finger">
    <visual>
      <origin rpy="0 0 0" xyz="0 -0.006  0"/>
      <geometry>
        <box size="0.04 0.012 0.025"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 -0.006  0"/>
      <geometry>
        <box size="0.04 0.012 0.025"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.012"/>
      <inertia
        ixx="0.000000732" ixy="0.0" ixz="0.0"
        iyy="0.000002118" iyz="0.0"
        izz="0.00000166"/>
    </inertial>
  </link>
  <joint name="gripper_to_right_finger_joint" type="prismatic">
    <parent link="gripper_base"/>
    <child link="right_finger"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <axis xyz="0 1 0"/>
    <limit effort="1000.0" lower="-0.11" upper="0" velocity="0.5"/>
  </joint>

  <link name="left_finger">
    <visual>
      <origin rpy="0 0 0" xyz="0 0.006 0"/>
      <geometry>
        <box size="0.04 0.012 0.025"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0.006 0"/>
      <geometry>
        <box size="0.04 0.012 0.025"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.012"/>
      <inertia
        ixx="0.000000732" ixy="0.0" ixz="0.0"
        iyy="0.000002118" iyz="0.0"
        izz="0.00000166"/>
    </inertial>
  </link>
  <joint name="gripper_to_left_finger_joint" type="prismatic">
    <parent link="gripper_base"/>
    <child link="left_finger"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <axis xyz="0 1 0"/>
    <limit effort="1000.0" lower="0" upper="0.11" velocity="0.5"/>
  </joint>

  <link name="camera_servo">
    <visual>
      <origin rpy="0 0 0" xyz="-0.01 0.015 0.008"/>
      <geometry>
        <box size="0.035 0.03 0.016"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="-0.01 0.015 0.008"/>
      <geometry>
        <box size="0.03 0.035 0.016"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.012"/>
      <inertia
        ixx="0.000000732" ixy="0.0" ixz="0.0"
        iyy="0.000002118" iyz="0.0"
        izz="0.00000166"/>
    </inertial>
  </link>
  <joint name="base_to_camera_servo_joint" type="fixed">
    <parent link="base_link"/>
    <child link="camera_servo"/>
    <origin rpy="0 0 0" xyz="0.21 0 0.27"/>
  </joint>

  <link name="camera_base">
    <visual>
      <origin rpy="0 0 0" xyz="0.02 0 -0.02"/>
      <geometry>
        <box size="0.004 0.04 0.04"/>
      </geometry>
    </visual>
    <collision>
      origin rpy="0 0 0" xyz="-0.01 0.015 0.01"/>
      <geometry>
        <box size="0.002 0.04 0.04"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.012"/>
      <inertia
        ixx="0.000000732" ixy="0.0" ixz="0.0"
        iyy="0.000002118" iyz="0.0"
        izz="0.00000166"/>
    </inertial>
  </link>
  <joint name="camera_servo_to_camera_base_joint" type="revolute">
    <parent link="camera_servo"/>
    <child link="camera_base"/>
    <origin rpy="0 1.57075 3.14159" xyz="0 0 0"/>
    <axis xyz="0 1 0"/>
    <limit effort="1000.0" lower="0.69" upper="3.14159" velocity="0.5"/>
  </joint>

  <link name="camera_center">
    <visual>
      <origin rpy="0 1.57075 0" xyz="0.01 0 0"/>
      <geometry>
        <cylinder length="0.02" radius="0.006"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="-0.01 0 0"/>
      <geometry>
        <cylinder length="0.02" radius="0.006"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.012"/>
      <inertia
        ixx="0.000000732" ixy="0.0" ixz="0.0"
        iyy="0.000002118" iyz="0.0"
        izz="0.00000166"/>
    </inertial>
  </link>
  <joint name="camera_base_to_camera_center_joint" type="fixed">
    <parent link="camera_base"/>
    <child link="camera_center"/>
    <origin rpy="0 0 0" xyz="0.02 0 -0.02"/>
  </joint>




</robot>
